src/common/core/utils.py

```diff
@@ -186,3 +195,26 @@ def using_database_replica(
         return manager
 
     return manager.db_manager(chosen_replica)
+
+
+if sys.version_info >= (3, 12):
+    # Already has the desired behavior; re-export for uniform imports.
+    TemporaryDirectory = tempfile.TemporaryDirectory
+else:
+    import contextlib
+    from typing import ContextManager, Generator
+
+    def TemporaryDirectory(
+        suffix: str | None = None,
+        prefix: str | None = None,
+        dir: str | None = None,
+        delete: bool = True,
+    ) -> ContextManager[str]:
+        if delete:
+            return tempfile.TemporaryDirectory(suffix, prefix, dir)
+
+        @contextlib.contextmanager
+        def _tmpdir() -> Generator[str, None, None]:
+            yield tempfile.mkdtemp(suffix, prefix, dir)
+
+        return _tmpdir()
```

Comment:
I think it's probably worth adding some more comments / documentation to this code to explain why it exists, and try to explain a little bit about how it works?